torch/csrc/autograd/variable.h

```diff
@@ -333,40 +303,82 @@ struct Variable::Impl : public at::TensorImpl {
 
   std::shared_ptr<Function> get_grad_accumulator();
   virtual std::shared_ptr<Function>& get_grad_fn() {
-    return grad_fn;
+    return grad_fn_;
+  }
+
+  virtual const Variable& base() const {
+    throw std::runtime_error("Can't get base of non-view Variable");
+  }
+
+  /// Sets the `requires_grad` property of `Variable`. This should be true for
+  /// leaf variables that want to accumulate gradients, and false for all other
+  /// variables.
+  void set_requires_grad(bool requires_grad) noexcept override {
+    requires_grad_ = requires_grad;
+  }
+
+  bool requires_grad() const noexcept override {
+    return requires_grad_ || grad_fn_ || (is_view_ && base().requires_grad());
```

Comment:
This is virtual now, so we don't need `is_view_` anymore. It was there only to be a faster alternative to virtual dispatch.

Reply:
Yeah but right now `base()` will throw an exception when called for a non-view. And if I change  `base()` to return `*this` instead, then `base().requires_grad()` will infinite loop. So unless I do a `dynamic_cast`, I'm not sure there's a better way to solve it right now. What do you think?

Reply:
Ugh... Ok, let's keep it.